tianshou/policy/imitation/cql.py

```diff
@@ -251,6 +255,23 @@ def learn(self, batch: RolloutBatchProtocol, *args: Any, **kwargs: Any) -> dict[
         ]:
             value.reshape(batch_size, self.num_repeat_actions, 1)
 
+        if self.calibrated:
```

Comment:
Note that with the current design, the user has to know to add returns to the batch under the exactly right field name, which means this code is fragile.

Instead, when `calibrated=True`, the returns should be added to the batch inside the policy class itself. Make sure to use the right `BatchProtocol` for expressing that, if needed (or make a new one, if there is no suitable).

Reply:
I agree that this is a fragile implementation, however the solution you are proposing comes at a high performance cost. Maybe adding some checks is the way to go for now?

Reply:
Why would there be a performance cost? You can add new values to the batch in `process_fn`, which takes the whole buffer. Am I missing something?

Adding checks is not the way to go here, we can't expect a Policy to only work if a user creates some external function that carefully processes the buffer

Reply:
`process_fn` is executed at every learning step that's where the performance cost comes from, compared to doing it only one time before starting training. I think this comes from the fundamental difference between offline and online RL, so maybe better to include a preprocess buffer step to the OfflineTrainer where this kind of one time modifications that are possible in offline RL are possible.

Reply:
Ah, I see. I want to adjust the control flow of update->preprocess->learn in the near future, in part to avoid such problems. Memoization would also help here... It can't be solved in the trainer, as the trainer doesn't know about the `calibrated` arg.

I suggest the following: 
1. Add a new parameter to CQLPolicy called `offline_mode: bool = True`. Document that with this, things like `process_fn` will only be called once.
2. Add hidden fields a la `_buffer_was_processed = False`
3. Use this field to make sure `process_fn` is called only once during training. You should also be able to reset the fields, maybe adding a `reset` method to the policy

It's not particularly clean, but should do the job until the general structure has been improved

Reply:
The execution of `process_fn` is outside the policy implementation, it is called in other places such as `collector`, so unless I'm missing something, I don't think it's advisable nor possible to specify how many times that it should be called. My idea is to add a `preprocess_buffer` to the policy, that if it exists in the policy implementation it will be called once in the OfflineTrianer.

Reply:
Ok, that's even better! Then it should be added to `BasePolicy`, where by default it just returns the unmodified buffer, and overwritten with `@overwrite` in CQLPolicy. The BaseTrainer can be adjusted to always call this function. Could you pls also add this logic to the Trainer's documentation?